### ISSUE TITLE
bug - error passed to callback goes to global unhandledRejection handler

### ISSUE THREAD

**Issue by ORESoftware** [octokit/octokit.js#1022]
Not sure why this happens, probably something to do with interop between promise API and vanilla callback API:

```js
const Octokit = require('@octokit/rest');

const github = new Octokit({
  // optional
  debug: false,
  // protocol: 'https',
  // host: 'api.github.com', // should be api.github.com for GitHub
  // pathPrefix: '', // for some GHEs; none for GitHub
  timeout: 5000,
  headers: {
    'user-agent': 'My-Cool-GitHub-App' // GitHub is happy with a unique user agent
  }
});

github.users.getOrgMemberships({state: 'active'}, (err: any, res: any) => {

  if (err) {
    return cb(err, null);
  }

 try{
    res = <Array<any>>JSON.parse(res);
    res = res.data || res;
    assert(Array.isArray(res), 'Github API response was not any array.');  // THROW HERE
 }
 catch(err){
   return cb(err);  // gets CAUGHT HERE
} 

  cb(null, res.map((item: any, index: number) => {
    const login = item.organization && item.organization.login;
    return String(login || ' *** unknown *** ').toUpperCase();
  }));

});
```

if we throw an error in where it says THROW ERROR, then the error hits the global handler:

```js
process.on('unhandledRejection', (r, p) => {
     //  r.message === 'Github API response was not any array.'
});
```

why is that?  Might be poor design TBH.

I know you are designing a promise-first API, but this still shouldn't happen even if it it is promise-first.

**Comment by gr2m** [comment]
where does `cb` come from in your code?

**Comment by gr2m** [comment]
Can you make a reproducable test case such as https://runkit.com/gr2m/octokit-rest-js-587? That’ll help us investigate and make sure that there is in fact a problem with the code.

**Comment by gr2m** [comment]
closing due to inactivity

**Comment by ORESoftware** [comment]
@gr2m please re-open, here is the runkit demo:
https://runkit.com/oresoftware/5bb40a46e4225d001251a39f

you asked:

> where does cb come from in your code?

it's now in the runkit demo, nothing fancy

the reason the `unhandledRejection` event occurs might be because of the way the `es6-promise` library works...but ultimately the error should not be an `unhandledRejection`, the error should be passed to the callback.

<img width="1134" alt="screenshot 2018-10-02 17 38 30" src="https://user-images.githubusercontent.com/11139560/46384316-153dba00-c66a-11e8-815e-e9d4a950aa0b.png">

In other words, if a callback is passed, the error should get passed to the callback, and no `unhandledRejection` should be raised.

**Comment by gr2m** [comment]
I’ll get back to this once I’m through with v16

**Comment by ORESoftware** [comment]
ok cool, i have seen this problem in other places, not just this library - it has to do with the es6-promise library

**Comment by gr2m** [comment]
We are removing callbacks from the API in v16, so I fear I won’t look into it after all I’m afraid

**Comment by gr2m** [comment]
obsolete via #1052

**Comment by ORESoftware** [comment]
thats good, although not that hard to support both, but the lib you were using before was no bueno.

any hope for typescript support? i tend to only use libs written in ts

I use this pattern in one my libs:

 ```js
class Foo {

  bar(v, cb){
     setTimeout(cb, 5);
  },

  barp(v){
    return new Promise((resolve, reject) => {
          this.bar(v, (err, val) => {
                 err ? reject(err) : resolve(val);
          });
     });
  }


```

with static typing and no operator overloading, best to just have to separare calls.
